src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+        /// have any information for this context it doesn't need to be delegated
+        /// </summary>
+        protected abstract Task<TResponse?> HandleInRazorAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// The delegated object to send to the <see cref="EndpointName"/>
+        /// </summary>
+        protected abstract TDelegatedParams CreateDelegatedParams(TRequest request, DocumentContext documentContext, Projection projection);
+
+        /// <summary>
+        /// The name of the endpoint to delegate to, from <see cref="LanguageServerConstants"/>
+        /// </summary>
+        protected abstract string EndpointName { get; }
+
+        /// <summary>
+        /// If the response needs to be remapped for any reasons, override to handle the remapping logic
+        /// </summary>
+        protected virtual Task<TResponse> RemapResponseAsync(TResponse delegatedResponse, RazorCodeDocument codeDocument, CancellationToken cancellationToken)
+            => Task.FromResult(delegatedResponse);
+
+        /// <summary>
+        /// Extracts information about the request and then calls <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>.
+        /// Children can override to provide any more custom logic needed before doing a normal delegated response pass with <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>
+        /// </summary>
+        public virtual async Task<TResponse?> Handle(TRequest request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentContext = await _documentContextFactory.TryCreateAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+            if (documentContext is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var absoluteIndex))
+            {
+                return null;
+            }
+
+            var projection = await _documentMappingService.GetProjectionAsync(documentContext, absoluteIndex, cancellationToken).ConfigureAwait(false);
+
+            return await GetDelegatedResponseAsync(request, documentContext, codeDocument, sourceText, absoluteIndex, projection, cancellationToken).ConfigureAwait(false);
```

Comment:
nit: for debuggability I typically like to have an intermediate assignee:

```suggestion
            var delegatedResponse = await GetDelegatedResponseAsync(request, documentContext, codeDocument, sourceText, absoluteIndex, projection, cancellationToken).ConfigureAwait(false);
            return delegatedResponse;
```

Reply:
Hmm, with the current approach we aren't able to combine both delegated data + non-delegated data. For instance, OnAUtoInsert takes both delegated information & our own information. Effectively we'll want a "Merge delegated + Razor info"

Reply:
We already have a remap after delegated gets called, maybe that's the appropriate place to do that?